{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: Could not connect to the Neo4j Database. See console for details.\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:806\u001b[0m, in \u001b[0;36mConnectionPool.acquire\u001b[1;34m(self, force_reset, can_overfill)\u001b[0m\n\u001b[0;32m    804\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    805\u001b[0m     \u001b[39m# Plan A: select a free connection from the pool\u001b[39;00m\n\u001b[1;32m--> 806\u001b[0m     cx \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_free_list\u001b[39m.\u001b[39;49mpopleft()\n\u001b[0;32m    807\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mIndexError\u001b[39;00m:\n",
      "\u001b[1;31mIndexError\u001b[0m: pop from an empty deque",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mConnectionRefusedError\u001b[0m                    Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\wiring.py:62\u001b[0m, in \u001b[0;36mWire.open\u001b[1;34m(cls, address, timeout, keep_alive, on_broken)\u001b[0m\n\u001b[0;32m     61\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 62\u001b[0m     s\u001b[39m.\u001b[39;49mconnect(address)\n\u001b[0;32m     63\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m error:\n",
      "\u001b[1;31mConnectionRefusedError\u001b[0m: [WinError 10061] No connection could be made because the target machine actively refused it",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mWireError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\bolt.py:355\u001b[0m, in \u001b[0;36mBolt.open\u001b[1;34m(cls, profile, user_agent, on_release, on_broken)\u001b[0m\n\u001b[0;32m    354\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 355\u001b[0m     wire \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39;49m\u001b[39m.\u001b[39;49m_connect(profile, on_broken\u001b[39m=\u001b[39;49mon_broken)\n\u001b[0;32m    356\u001b[0m     protocol_version \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39m\u001b[39m.\u001b[39m_handshake(wire)\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\bolt.py:369\u001b[0m, in \u001b[0;36mBolt._connect\u001b[1;34m(cls, profile, on_broken)\u001b[0m\n\u001b[0;32m    368\u001b[0m log\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39m[#\u001b[39m\u001b[39m%04X\u001b[39;00m\u001b[39m] C: (Dialing <\u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m>)\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m0\u001b[39m, profile\u001b[39m.\u001b[39maddress)\n\u001b[1;32m--> 369\u001b[0m wire \u001b[39m=\u001b[39m Wire\u001b[39m.\u001b[39;49mopen(profile\u001b[39m.\u001b[39;49maddress, keep_alive\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, on_broken\u001b[39m=\u001b[39;49mon_broken)\n\u001b[0;32m    370\u001b[0m local_port \u001b[39m=\u001b[39m wire\u001b[39m.\u001b[39mlocal_address\u001b[39m.\u001b[39mport_number\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\wiring.py:64\u001b[0m, in \u001b[0;36mWire.open\u001b[1;34m(cls, address, timeout, keep_alive, on_broken)\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mIOError\u001b[39;00m, \u001b[39mOSError\u001b[39;00m) \u001b[39mas\u001b[39;00m error:\n\u001b[1;32m---> 64\u001b[0m     raise_from(WireError(\u001b[39m\"\u001b[39;49m\u001b[39mCannot connect to \u001b[39;49m\u001b[39m%r\u001b[39;49;00m\u001b[39m\"\u001b[39;49m \u001b[39m%\u001b[39;49m (address,)), error)\n\u001b[0;32m     65\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mcls\u001b[39m(s, on_broken\u001b[39m=\u001b[39mon_broken)\n",
      "File \u001b[1;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[1;34m(value, from_value)\u001b[0m\n",
      "\u001b[1;31mWireError\u001b[0m: Cannot connect to IPv4Address(('localhost', 7687))",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mConnectionUnavailable\u001b[0m                     Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[1], line 31\u001b[0m\n\u001b[0;32m     30\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m---> 31\u001b[0m     graph \u001b[39m=\u001b[39m Graph(\u001b[39m'\u001b[39;49m\u001b[39mbolt://localhost:\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m+\u001b[39;49mport, auth\u001b[39m=\u001b[39;49m(user, pswd))\n\u001b[0;32m     32\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mSUCCESS: Connected to the Neo4j Database.\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\database.py:288\u001b[0m, in \u001b[0;36mGraph.__init__\u001b[1;34m(self, profile, name, **settings)\u001b[0m\n\u001b[0;32m    287\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, profile\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, name\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39msettings):\n\u001b[1;32m--> 288\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mservice \u001b[39m=\u001b[39m GraphService(profile, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39msettings)\n\u001b[0;32m    289\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__name__\u001b[39m \u001b[39m=\u001b[39m name\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\database.py:119\u001b[0m, in \u001b[0;36mGraphService.__init__\u001b[1;34m(self, profile, **settings)\u001b[0m\n\u001b[0;32m    118\u001b[0m     connector_settings[\u001b[39m\"\u001b[39m\u001b[39minit_size\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n\u001b[1;32m--> 119\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_connector \u001b[39m=\u001b[39m Connector(profile, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mconnector_settings)\n\u001b[0;32m    120\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_graphs \u001b[39m=\u001b[39m {}\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:960\u001b[0m, in \u001b[0;36mConnector.__init__\u001b[1;34m(self, profile, user_agent, init_size, max_size, max_age, routing_refresh_ttl)\u001b[0m\n\u001b[0;32m    959\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_router \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m--> 960\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_add_pools(\u001b[39m*\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_initial_routers)\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:982\u001b[0m, in \u001b[0;36mConnector._add_pools\u001b[1;34m(self, *profiles)\u001b[0m\n\u001b[0;32m    981\u001b[0m log\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mAdding connection pool for profile \u001b[39m\u001b[39m%r\u001b[39;00m\u001b[39m\"\u001b[39m, profile)\n\u001b[1;32m--> 982\u001b[0m pool \u001b[39m=\u001b[39m ConnectionPool\u001b[39m.\u001b[39;49mopen(\n\u001b[0;32m    983\u001b[0m     profile,\n\u001b[0;32m    984\u001b[0m     user_agent\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_user_agent,\n\u001b[0;32m    985\u001b[0m     init_size\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_init_size,\n\u001b[0;32m    986\u001b[0m     max_size\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_max_size,\n\u001b[0;32m    987\u001b[0m     max_age\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_max_age,\n\u001b[0;32m    988\u001b[0m     on_broken\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_on_broken)\n\u001b[0;32m    989\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_pools[profile] \u001b[39m=\u001b[39m pool\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:649\u001b[0m, in \u001b[0;36mConnectionPool.open\u001b[1;34m(cls, profile, user_agent, init_size, max_size, max_age, on_broken)\u001b[0m\n\u001b[0;32m    648\u001b[0m pool \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39m(profile, user_agent, max_size, max_age, on_broken)\n\u001b[1;32m--> 649\u001b[0m seeds \u001b[39m=\u001b[39m [pool\u001b[39m.\u001b[39macquire() \u001b[39mfor\u001b[39;00m _ \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(init_size \u001b[39mor\u001b[39;00m \u001b[39mcls\u001b[39m\u001b[39m.\u001b[39mdefault_init_size)]\n\u001b[0;32m    650\u001b[0m \u001b[39mfor\u001b[39;00m seed \u001b[39min\u001b[39;00m seeds:\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:649\u001b[0m, in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m    648\u001b[0m pool \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39m(profile, user_agent, max_size, max_age, on_broken)\n\u001b[1;32m--> 649\u001b[0m seeds \u001b[39m=\u001b[39m [pool\u001b[39m.\u001b[39;49macquire() \u001b[39mfor\u001b[39;00m _ \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(init_size \u001b[39mor\u001b[39;00m \u001b[39mcls\u001b[39m\u001b[39m.\u001b[39mdefault_init_size)]\n\u001b[0;32m    650\u001b[0m \u001b[39mfor\u001b[39;00m seed \u001b[39min\u001b[39;00m seeds:\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:813\u001b[0m, in \u001b[0;36mConnectionPool.acquire\u001b[1;34m(self, force_reset, can_overfill)\u001b[0m\n\u001b[0;32m    808\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_has_capacity() \u001b[39mor\u001b[39;00m can_overfill:\n\u001b[0;32m    809\u001b[0m     \u001b[39m# Plan B: if the pool isn't full, open\u001b[39;00m\n\u001b[0;32m    810\u001b[0m     \u001b[39m# a new connection. This may raise a\u001b[39;00m\n\u001b[0;32m    811\u001b[0m     \u001b[39m# ConnectionUnavailable exception, which\u001b[39;00m\n\u001b[0;32m    812\u001b[0m     \u001b[39m# should bubble up to the caller.\u001b[39;00m\n\u001b[1;32m--> 813\u001b[0m     cx \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_connect()\n\u001b[0;32m    814\u001b[0m     \u001b[39mif\u001b[39;00m cx\u001b[39m.\u001b[39msupports_multi():\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:764\u001b[0m, in \u001b[0;36mConnectionPool._connect\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    762\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\" Open and return a new connection.\u001b[39;00m\n\u001b[0;32m    763\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m--> 764\u001b[0m cx \u001b[39m=\u001b[39m Connection\u001b[39m.\u001b[39;49mopen(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mprofile, user_agent\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49muser_agent,\n\u001b[0;32m    765\u001b[0m                      on_release\u001b[39m=\u001b[39;49m\u001b[39mlambda\u001b[39;49;00m c: \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrelease(c),\n\u001b[0;32m    766\u001b[0m                      on_broken\u001b[39m=\u001b[39;49m\u001b[39mlambda\u001b[39;49;00m msg: \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__on_broken(msg))\n\u001b[0;32m    767\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_server_agent \u001b[39m=\u001b[39m cx\u001b[39m.\u001b[39mserver_agent\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\__init__.py:174\u001b[0m, in \u001b[0;36mConnection.open\u001b[1;34m(cls, profile, user_agent, on_release, on_broken)\u001b[0m\n\u001b[0;32m    173\u001b[0m     \u001b[39mfrom\u001b[39;00m \u001b[39mpy2neo\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mclient\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mbolt\u001b[39;00m \u001b[39mimport\u001b[39;00m Bolt\n\u001b[1;32m--> 174\u001b[0m     \u001b[39mreturn\u001b[39;00m Bolt\u001b[39m.\u001b[39;49mopen(profile, user_agent\u001b[39m=\u001b[39;49muser_agent,\n\u001b[0;32m    175\u001b[0m                      on_release\u001b[39m=\u001b[39;49mon_release, on_broken\u001b[39m=\u001b[39;49mon_broken)\n\u001b[0;32m    176\u001b[0m \u001b[39melif\u001b[39;00m profile\u001b[39m.\u001b[39mprotocol \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mhttp\u001b[39m\u001b[39m\"\u001b[39m:\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\py2neo\\client\\bolt.py:364\u001b[0m, in \u001b[0;36mBolt.open\u001b[1;34m(cls, profile, user_agent, on_release, on_broken)\u001b[0m\n\u001b[0;32m    363\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mTypeError\u001b[39;00m, WireError) \u001b[39mas\u001b[39;00m error:\n\u001b[1;32m--> 364\u001b[0m     raise_from(ConnectionUnavailable(\u001b[39m\"\u001b[39;49m\u001b[39mCannot open connection to \u001b[39;49m\u001b[39m%r\u001b[39;49;00m\u001b[39m\"\u001b[39;49m \u001b[39m%\u001b[39;49m profile), error)\n",
      "File \u001b[1;32m<string>:3\u001b[0m, in \u001b[0;36mraise_from\u001b[1;34m(value, from_value)\u001b[0m\n",
      "\u001b[1;31mConnectionUnavailable\u001b[0m: Cannot open connection to ConnectionProfile('bolt://localhost:7687')",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mSystemExit\u001b[0m                                Traceback (most recent call last)",
      "    \u001b[1;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "Cell \u001b[1;32mIn[1], line 35\u001b[0m\n\u001b[0;32m     34\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39mERROR: Could not connect to the Neo4j Database. See console for details.\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m---> 35\u001b[0m \u001b[39mraise\u001b[39;00m \u001b[39mSystemExit\u001b[39;00m(e)\n",
      "\u001b[1;31mSystemExit\u001b[0m: Cannot open connection to ConnectionProfile('bolt://localhost:7687')",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "    \u001b[1;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py:2047\u001b[0m, in \u001b[0;36mInteractiveShell.showtraceback\u001b[1;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001b[0m\n\u001b[0;32m   2044\u001b[0m \u001b[39mif\u001b[39;00m exception_only:\n\u001b[0;32m   2045\u001b[0m     stb \u001b[39m=\u001b[39m [\u001b[39m'\u001b[39m\u001b[39mAn exception has occurred, use \u001b[39m\u001b[39m%\u001b[39m\u001b[39mtb to see \u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m   2046\u001b[0m            \u001b[39m'\u001b[39m\u001b[39mthe full traceback.\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m]\n\u001b[1;32m-> 2047\u001b[0m     stb\u001b[39m.\u001b[39mextend(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mInteractiveTB\u001b[39m.\u001b[39;49mget_exception_only(etype,\n\u001b[0;32m   2048\u001b[0m                                                      value))\n\u001b[0;32m   2049\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   2050\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m   2051\u001b[0m         \u001b[39m# Exception classes can customise their traceback - we\u001b[39;00m\n\u001b[0;32m   2052\u001b[0m         \u001b[39m# use this in IPython.parallel for exceptions occurring\u001b[39;00m\n\u001b[0;32m   2053\u001b[0m         \u001b[39m# in the engines. This should return a list of strings.\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:585\u001b[0m, in \u001b[0;36mListTB.get_exception_only\u001b[1;34m(self, etype, value)\u001b[0m\n\u001b[0;32m    577\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_exception_only\u001b[39m(\u001b[39mself\u001b[39m, etype, value):\n\u001b[0;32m    578\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Only print the exception type and message, without a traceback.\u001b[39;00m\n\u001b[0;32m    579\u001b[0m \n\u001b[0;32m    580\u001b[0m \u001b[39m    Parameters\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    583\u001b[0m \u001b[39m    value : exception value\u001b[39;00m\n\u001b[0;32m    584\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 585\u001b[0m     \u001b[39mreturn\u001b[39;00m ListTB\u001b[39m.\u001b[39;49mstructured_traceback(\u001b[39mself\u001b[39;49m, etype, value)\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:452\u001b[0m, in \u001b[0;36mListTB.structured_traceback\u001b[1;34m(self, etype, evalue, etb, tb_offset, context)\u001b[0m\n\u001b[0;32m    449\u001b[0m     chained_exc_ids\u001b[39m.\u001b[39madd(\u001b[39mid\u001b[39m(exception[\u001b[39m1\u001b[39m]))\n\u001b[0;32m    450\u001b[0m     chained_exceptions_tb_offset \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m\n\u001b[0;32m    451\u001b[0m     out_list \u001b[39m=\u001b[39m (\n\u001b[1;32m--> 452\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mstructured_traceback(\n\u001b[0;32m    453\u001b[0m             etype, evalue, (etb, chained_exc_ids),\n\u001b[0;32m    454\u001b[0m             chained_exceptions_tb_offset, context)\n\u001b[0;32m    455\u001b[0m         \u001b[39m+\u001b[39m chained_exception_message\n\u001b[0;32m    456\u001b[0m         \u001b[39m+\u001b[39m out_list)\n\u001b[0;32m    458\u001b[0m \u001b[39mreturn\u001b[39;00m out_list\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:1118\u001b[0m, in \u001b[0;36mAutoFormattedTB.structured_traceback\u001b[1;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[0;32m   1116\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   1117\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtb \u001b[39m=\u001b[39m tb\n\u001b[1;32m-> 1118\u001b[0m \u001b[39mreturn\u001b[39;00m FormattedTB\u001b[39m.\u001b[39;49mstructured_traceback(\n\u001b[0;32m   1119\u001b[0m     \u001b[39mself\u001b[39;49m, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:1012\u001b[0m, in \u001b[0;36mFormattedTB.structured_traceback\u001b[1;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[0;32m   1009\u001b[0m mode \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmode\n\u001b[0;32m   1010\u001b[0m \u001b[39mif\u001b[39;00m mode \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mverbose_modes:\n\u001b[0;32m   1011\u001b[0m     \u001b[39m# Verbose modes need a full traceback\u001b[39;00m\n\u001b[1;32m-> 1012\u001b[0m     \u001b[39mreturn\u001b[39;00m VerboseTB\u001b[39m.\u001b[39;49mstructured_traceback(\n\u001b[0;32m   1013\u001b[0m         \u001b[39mself\u001b[39;49m, etype, value, tb, tb_offset, number_of_lines_of_context\n\u001b[0;32m   1014\u001b[0m     )\n\u001b[0;32m   1015\u001b[0m \u001b[39melif\u001b[39;00m mode \u001b[39m==\u001b[39m \u001b[39m'\u001b[39m\u001b[39mMinimal\u001b[39m\u001b[39m'\u001b[39m:\n\u001b[0;32m   1016\u001b[0m     \u001b[39mreturn\u001b[39;00m ListTB\u001b[39m.\u001b[39mget_exception_only(\u001b[39mself\u001b[39m, etype, value)\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:865\u001b[0m, in \u001b[0;36mVerboseTB.structured_traceback\u001b[1;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[0;32m    856\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mstructured_traceback\u001b[39m(\n\u001b[0;32m    857\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m    858\u001b[0m     etype: \u001b[39mtype\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    862\u001b[0m     number_of_lines_of_context: \u001b[39mint\u001b[39m \u001b[39m=\u001b[39m \u001b[39m5\u001b[39m,\n\u001b[0;32m    863\u001b[0m ):\n\u001b[0;32m    864\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Return a nice text document describing the traceback.\"\"\"\u001b[39;00m\n\u001b[1;32m--> 865\u001b[0m     formatted_exception \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mformat_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n\u001b[0;32m    866\u001b[0m                                                            tb_offset)\n\u001b[0;32m    868\u001b[0m     colors \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mColors  \u001b[39m# just a shorthand + quicker name lookup\u001b[39;00m\n\u001b[0;32m    869\u001b[0m     colorsnormal \u001b[39m=\u001b[39m colors\u001b[39m.\u001b[39mNormal  \u001b[39m# used a lot\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:799\u001b[0m, in \u001b[0;36mVerboseTB.format_exception_as_a_whole\u001b[1;34m(self, etype, evalue, etb, number_of_lines_of_context, tb_offset)\u001b[0m\n\u001b[0;32m    796\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39misinstance\u001b[39m(tb_offset, \u001b[39mint\u001b[39m)\n\u001b[0;32m    797\u001b[0m head \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprepare_header(etype, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mlong_header)\n\u001b[0;32m    798\u001b[0m records \u001b[39m=\u001b[39m (\n\u001b[1;32m--> 799\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mget_records(etb, number_of_lines_of_context, tb_offset) \u001b[39mif\u001b[39;00m etb \u001b[39melse\u001b[39;00m []\n\u001b[0;32m    800\u001b[0m )\n\u001b[0;32m    802\u001b[0m frames \u001b[39m=\u001b[39m []\n\u001b[0;32m    803\u001b[0m skipped \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\ultratb.py:854\u001b[0m, in \u001b[0;36mVerboseTB.get_records\u001b[1;34m(self, etb, number_of_lines_of_context, tb_offset)\u001b[0m\n\u001b[0;32m    848\u001b[0m     formatter \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m\n\u001b[0;32m    849\u001b[0m options \u001b[39m=\u001b[39m stack_data\u001b[39m.\u001b[39mOptions(\n\u001b[0;32m    850\u001b[0m     before\u001b[39m=\u001b[39mbefore,\n\u001b[0;32m    851\u001b[0m     after\u001b[39m=\u001b[39mafter,\n\u001b[0;32m    852\u001b[0m     pygments_formatter\u001b[39m=\u001b[39mformatter,\n\u001b[0;32m    853\u001b[0m )\n\u001b[1;32m--> 854\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mlist\u001b[39;49m(stack_data\u001b[39m.\u001b[39;49mFrameInfo\u001b[39m.\u001b[39;49mstack_data(etb, options\u001b[39m=\u001b[39;49moptions))[tb_offset:]\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\stack_data\\core.py:578\u001b[0m, in \u001b[0;36mFrameInfo.stack_data\u001b[1;34m(cls, frame_or_tb, options, collapse_repeated_frames)\u001b[0m\n\u001b[0;32m    562\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[0;32m    563\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mstack_data\u001b[39m(\n\u001b[0;32m    564\u001b[0m         \u001b[39mcls\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    568\u001b[0m         collapse_repeated_frames: \u001b[39mbool\u001b[39m \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[0;32m    569\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m Iterator[Union[\u001b[39m'\u001b[39m\u001b[39mFrameInfo\u001b[39m\u001b[39m'\u001b[39m, RepeatedFrames]]:\n\u001b[0;32m    570\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m    571\u001b[0m \u001b[39m    An iterator of FrameInfo and RepeatedFrames objects representing\u001b[39;00m\n\u001b[0;32m    572\u001b[0m \u001b[39m    a full traceback or stack. Similar consecutive frames are collapsed into RepeatedFrames\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    576\u001b[0m \u001b[39m    and optionally an Options object to configure.\u001b[39;00m\n\u001b[0;32m    577\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 578\u001b[0m     stack \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39;49m(iter_stack(frame_or_tb))\n\u001b[0;32m    580\u001b[0m     \u001b[39m# Reverse the stack from a frame so that it's in the same order\u001b[39;00m\n\u001b[0;32m    581\u001b[0m     \u001b[39m# as the order from a traceback, which is the order of a printed\u001b[39;00m\n\u001b[0;32m    582\u001b[0m     \u001b[39m# traceback when read top to bottom (most recent call last)\u001b[39;00m\n\u001b[0;32m    583\u001b[0m     \u001b[39mif\u001b[39;00m is_frame(frame_or_tb):\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\stack_data\\utils.py:97\u001b[0m, in \u001b[0;36miter_stack\u001b[1;34m(frame_or_tb)\u001b[0m\n\u001b[0;32m     95\u001b[0m \u001b[39mwhile\u001b[39;00m frame_or_tb:\n\u001b[0;32m     96\u001b[0m     \u001b[39myield\u001b[39;00m frame_or_tb\n\u001b[1;32m---> 97\u001b[0m     \u001b[39mif\u001b[39;00m is_frame(frame_or_tb):\n\u001b[0;32m     98\u001b[0m         frame_or_tb \u001b[39m=\u001b[39m frame_or_tb\u001b[39m.\u001b[39mf_back\n\u001b[0;32m     99\u001b[0m     \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\stack_data\\utils.py:90\u001b[0m, in \u001b[0;36mis_frame\u001b[1;34m(frame_or_tb)\u001b[0m\n\u001b[0;32m     89\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mis_frame\u001b[39m(frame_or_tb: Union[FrameType, TracebackType]) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mbool\u001b[39m:\n\u001b[1;32m---> 90\u001b[0m     assert_(\u001b[39misinstance\u001b[39;49m(frame_or_tb, (types\u001b[39m.\u001b[39;49mFrameType, types\u001b[39m.\u001b[39;49mTracebackType)))\n\u001b[0;32m     91\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39misinstance\u001b[39m(frame_or_tb, (types\u001b[39m.\u001b[39mFrameType,))\n",
      "File \u001b[1;32mc:\\Users\\DKDAVA\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\stack_data\\utils.py:176\u001b[0m, in \u001b[0;36massert_\u001b[1;34m(condition, error)\u001b[0m\n\u001b[0;32m    174\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(error, \u001b[39mstr\u001b[39m):\n\u001b[0;32m    175\u001b[0m     error \u001b[39m=\u001b[39m \u001b[39mAssertionError\u001b[39;00m(error)\n\u001b[1;32m--> 176\u001b[0m \u001b[39mraise\u001b[39;00m error\n",
      "\u001b[1;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from py2neo import Graph\n",
    "import os\n",
    "\n",
    "# assign directory\n",
    "directory = 'C:/Users/DKDAVA/repos/AdAM_dataset/phuse-scripts/data/adam/TDF_ADaM_v1.0/'\n",
    "# giving file extension\n",
    "ext = ('.xpt')\n",
    "# create empty list to dataset names\n",
    "dataset_list = list()\n",
    "\n",
    "# iterate over all files with SAS extension\n",
    "for files in os.listdir(directory):\n",
    "    if files.endswith(ext):\n",
    "        m=files.split(sep='.')[0]\n",
    "        if type(m) != 'NoneType':\n",
    "            # append dataset name to list\n",
    "            dataset_list.append(m)\n",
    "        # assign datasetname to dataframe name (not a good practice)\n",
    "        globals()[m] = pd.read_sas(directory+files,format='xport') \n",
    "    else:\n",
    "        continue\n",
    "\n",
    "# neo4j connection\n",
    "port = \"7687\"\n",
    "user = \"neo4j\"\n",
    "pswd = \"adam\"\n",
    "\n",
    "# Make sure the database is started first, otherwise attempt to connect will fail\n",
    "try:\n",
    "    graph = Graph('bolt://localhost:'+port, auth=(user, pswd))\n",
    "    print('SUCCESS: Connected to the Neo4j Database.')\n",
    "except Exception as e:\n",
    "    print('ERROR: Could not connect to the Neo4j Database. See console for details.')\n",
    "    raise SystemExit(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating patient nodes\n",
    "for row in adsl.itertuples():\n",
    "    graph.run('''\n",
    "      CREATE (:Patient {USUBJID: $USUBJID, AGE: $AGE, ARM:$ARM, SEX: $SEX, BMI: $BMI})\n",
    "    ''', parameters = {'USUBJID': row.USUBJID, 'AGE': row.AGE, 'ARM': row.ARM, 'SEX': row.SEX, 'BMI': row.BMIBL})\n",
    "\n",
    "# Creating treatment Nodes\n",
    "treatment = adsl.groupby('ARM')['USUBJID'].count()\n",
    "for i, v in treatment.items():\n",
    "    graph.run('''\n",
    "      CREATE (t:Treatment {Name: $ARM})\n",
    "    ''', parameters = {'ARM': i})\n",
    "    \n",
    "# creating patient-treatment relationship\n",
    "for row in adsl.itertuples():\n",
    "    graph.run(f'''\n",
    "    MATCH (p:Patient {{USUBJID: '{str(row.USUBJID,'utf-8')}'}}), (t:Treatment {{Name:'{str(row.ARM,'utf-8')}'}}) \n",
    "    MERGE (p)-[r:WAS_TREATED]->(t)\n",
    "    SET r.Dose={row.TRT01PN}\n",
    "    ''')\n",
    "# Creating Adverse Event Nodes\n",
    "adverse = adae.groupby('AETERM')['USUBJID'].count()\n",
    "for i, v in adverse.items():\n",
    "    graph.run('''\n",
    "      CREATE (ae:AdverseEvent {Term: $Term})\n",
    "    ''', parameters = {'Term': i})\n",
    "# Creating patient-adverseevent relationship\n",
    "for row in adae.itertuples():\n",
    "    graph.run(f'''\n",
    "    MATCH (p:Patient {{USUBJID: $USUBJID}}), (ae:AdverseEvent {{Term: $Term}}) \n",
    "    MERGE (p)-[r:HAD_ADVERSE_EVENT]->(ae)\n",
    "    SET r={{Severity: $Severity, Type: $Type}}\n",
    "    ''', parameters = {'USUBJID': row.USUBJID,'Term': row.AETERM, 'Severity':row.AESEV, 'Type':row.AEBODSYS})\n",
    "# Creating Visit Nodes\n",
    "visits = adlbc.groupby('VISIT')['USUBJID'].count()\n",
    "for i, v in visits.items():\n",
    "    graph.run('''\n",
    "      CREATE (v:Visit {Name: $Name})\n",
    "    ''', parameters = {'Name': str(i,'utf-8')})  \n",
    "# Creating patient-visit relationships\n",
    "for row in adlbc.itertuples():\n",
    "  graph.run(f'''\n",
    "  MATCH (p:Patient {{USUBJID: $USUBJID}}), (v:Visit {{Name: $Visit}}) \n",
    "  MERGE (p)-[:HAD_VISIT]->(v)\n",
    "  ''', parameters = {'USUBJID': row.USUBJID,'Visit': row.VISIT})\n",
    "# creating chemistry laboratory parameters\n",
    "for row in adlbc.itertuples():\n",
    "  graph.run(f'''\n",
    "  CREATE (p:Parameter {{USUBJID: $USUBJID, VISIT: $VISIT, Laboratory: $PARCAT1, Parameter: $PARAM, Value: $AVAL, Reference: $Reference, Dataset: 'adlbc'}})\n",
    "  ''', parameters = {'USUBJID':row.USUBJID, 'VISIT':row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL, 'Reference':row.LBNRIND, 'PARCAT1':row.PARCAT1})\n",
    "# creating hematology parameters\n",
    "for row in adlbh.itertuples():\n",
    "  graph.run(f'''\n",
    "  CREATE (p:Parameter {{USUBJID: $USUBJID, VISIT: $VISIT, Laboratory: $PARCAT1, Parameter: $PARAM, Value: $AVAL, Reference: $Reference, Dataset: 'adlbh'}})\n",
    "  ''', parameters = {'USUBJID':row.USUBJID, 'VISIT':row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL, 'Reference':row.LBNRIND, 'PARCAT1':row.PARCAT1})\n",
    "\n",
    "# creating remaining visit nodes\n",
    "visitshem = adlbh.groupby('VISIT')['USUBJID'].count()\n",
    "for i, v in visitshem.items():\n",
    "    graph.run('''\n",
    "      MERGE (:Visit {Name: $Name})\n",
    "    ''', parameters = {'Name': str(i,'utf-8')})\n",
    "# creating chemistry-patient-visit relationships\n",
    "for row in adlbc.itertuples():\n",
    "  graph.run(f'''\n",
    "  MATCH (p:Patient {{USUBJID: $USUBJID}}), (param:Parameter {{USUBJID: $USUBJID,Laboratory: $PARCAT1,Parameter: $PARAM, Value: $AVAL, Reference: $Reference, Dataset: 'adlbc'}}), (v:Visit {{Name: $Visit}})\n",
    "  WHERE p.USUBJID=param.USUBJID AND param.VISIT=$Visit\n",
    "  CREATE (p)-[lb:MEASURED_LABPARAMETER]->(param)<-[:MEASURED_IN_VISIT]-(v)\n",
    "  SET lb={{ChangeFromBaseline: $ChangeFromBaseline}}\n",
    "  ''', parameters = {'USUBJID': row.USUBJID,'Visit': row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL, 'Reference':row.LBNRIND, 'PARCAT1':row.PARCAT1,'ChangeFromBaseline':row.CHG})\n",
    "  # creating hematology-patient-visit relationships\n",
    "for row in adlbh.itertuples():\n",
    "  graph.run(f'''\n",
    "  MATCH (p:Patient {{USUBJID: $USUBJID}}), (param:Parameter {{USUBJID: $USUBJID, VISIT: $Visit, Laboratory: $PARCAT1,Parameter: $PARAM, Value: $AVAL, Reference: $Reference, Dataset: 'adlbh'}}), (v:Visit {{Name: $Visit}})\n",
    "  WHERE p.USUBJID=param.USUBJID AND param.VISIT=$Visit\n",
    "  CREATE (p)-[lb:MEASURED_LABPARAMETER]->(param)<-[:MEASURED_IN_VISIT]-(v)\n",
    "  SET lb={{ChangeFromBaseline: $ChangeFromBaseline}}\n",
    "  ''', parameters = {'USUBJID': row.USUBJID,'Visit': row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL, 'Reference':row.LBNRIND, 'PARCAT1':row.PARCAT1,'ChangeFromBaseline':row.CHG})\n",
    "# creating remaining visit nodes\n",
    "visitsvs = advs.groupby('VISIT')['USUBJID'].count()\n",
    "for i, v in visitsvs.items():\n",
    "    graph.run('''\n",
    "      MERGE (:Visit {Name: $Name})\n",
    "    ''', parameters = {'Name': str(i,'utf-8')})\n",
    "# creating vital sign nodes\n",
    "for row in advs.itertuples():\n",
    "  graph.run(f'''\n",
    "  CREATE (vs:VitalSign {{USUBJID: $USUBJID, VISIT: $VISIT, Laboratory: 'VS', Parameter: $PARAM, Value: $AVAL, Reference: '', Dataset: 'adlvs'}})\n",
    "  ''', parameters = {'USUBJID':row.USUBJID, 'VISIT':row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL})\n",
    "# creating vitalsigns-patient-visit relationships\n",
    "for row in advs.itertuples():\n",
    "  graph.run(f'''\n",
    "  MATCH (p:Patient {{USUBJID: $USUBJID}}), (param:VitalSign {{USUBJID: $USUBJID, VISIT: $Visit, Laboratory: 'VS', Parameter: $PARAM, Value: $AVAL, Reference: '', Dataset: 'adlvs'}}), (v:Visit {{Name: $Visit}})\n",
    "  WHERE p.USUBJID=param.USUBJID AND param.VISIT=$Visit\n",
    "  CREATE (p)-[vs:MEASURED_VITALSIGN]->(param)<-[:MEASURED_IN_VISIT]-(v)\n",
    "  SET vs={{ChangeFromBaseline: $ChangeFromBaseline}}\n",
    "  ''', parameters = {'USUBJID': row.USUBJID,'Visit': row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL, 'ChangeFromBaseline':row.CHG})\n",
    "  # creating remaining visit nodes\n",
    "visitsendpoint = adadas.groupby('VISIT')['USUBJID'].count()\n",
    "for i, v in visitsendpoint.items():\n",
    "    graph.run('''\n",
    "      MERGE (:Visit {Name: $Name})\n",
    "    ''', parameters = {'Name': str(i,'utf-8')})\n",
    "# creating ADAS nodes\n",
    "for row in adadas.itertuples():\n",
    "  graph.run(f'''\n",
    "  CREATE (e:Endpoint {{USUBJID: $USUBJID, VISIT: $VISIT, EndpointName: 'ADAS-Cog', Parameter: $PARAM, Value: $AVAL, Reference: '', Dataset: 'adadas'}})\n",
    "  ''', parameters = {'USUBJID':row.USUBJID, 'VISIT':row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL})\n",
    "# creating ADAS-patient-visit relationships\n",
    "for row in adadas.itertuples():\n",
    "  graph.run(f'''\n",
    "  MATCH (p:Patient {{USUBJID: $USUBJID}}), (end:Endpoint {{USUBJID: $USUBJID, VISIT: $Visit, EndpointName: 'ADAS-Cog', Parameter: $PARAM, Value: $AVAL, Reference: '', Dataset: 'adadas'}}), (v:Visit {{Name: $Visit}})\n",
    "  WHERE p.USUBJID=end.USUBJID AND end.VISIT=$Visit\n",
    "  CREATE (p)-[endrel:ASSESSED_ENDPOINT]->(end)<-[:MEASURED_IN_VISIT]-(v)\n",
    "  SET endrel={{ChangeFromBaseline: $ChangeFromBaseline}}\n",
    "  ''', parameters = {'USUBJID': row.USUBJID,'Visit': row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL, 'ChangeFromBaseline':row.CHG})\n",
    "# creating remaining visit nodes\n",
    "visitsendpoint = adcibc.groupby('VISIT')['USUBJID'].count()\n",
    "for i, v in visitsendpoint.items():\n",
    "    graph.run('''\n",
    "      MERGE (:Visit {Name: $Name})\n",
    "    ''', parameters = {'Name': str(i,'utf-8')})\n",
    "# creating CIBC nodes\n",
    "for row in adcibc.itertuples():\n",
    "  graph.run(f'''\n",
    "  CREATE (e:Endpoint {{USUBJID: $USUBJID, VISIT: $VISIT, EndpointName: 'CIBC Score', Parameter: $PARAM, Value: $AVAL, Reference: '', Dataset: 'adcibc'}})\n",
    "  ''', parameters = {'USUBJID':row.USUBJID, 'VISIT':row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL})\n",
    "for row in adcibc.itertuples():\n",
    "  graph.run(f'''\n",
    "  MATCH (p:Patient {{USUBJID: $USUBJID}}), (end:Endpoint {{USUBJID: $USUBJID, VISIT: $Visit, EndpointName: 'CIBC Score', Parameter: $PARAM, Value: $AVAL, Reference: '', Dataset: 'adcibc'}}), (v:Visit {{Name: $Visit}})\n",
    "  WHERE p.USUBJID=end.USUBJID AND end.VISIT=$Visit\n",
    "  CREATE (p)-[endrel:ASSESSED_ENDPOINT]->(end)<-[:MEASURED_IN_VISIT]-(v)\n",
    "  SET endrel={{ChangeFromBaseline:0}}\n",
    "  ''', parameters = {'USUBJID': row.USUBJID,'Visit': row.VISIT, 'PARAM':row.PARAM, 'AVAL': row.AVAL})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a5ebd95462cb0493b7fda28be48b3f2c90f6504a5473ac8f1bce40b27ca1e7d0"
  },
  "kernelspec": {
   "display_name": "Python 3.10.0 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
